# Assignment: Data Wrangling
## `! git clone https://github.com/DS3001/wrangling`
## Do Q2, and one of Q1 or Q3.

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?

  This paper dives into the process of 'data tidying', the idea of making sure data is clean, easy-to-manipulate, model, and visualize through following a specific structure/framework. Each row is an observation, each column is a variable, and each type of observational unit is a table.

  2. Read the introduction. What is the "tidy data standard" intended to accomplish?

  The "tidy data standard" aims to simplify initial data exploration and analysis, along with the development of data analysis tools that work well together.

  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."

  The sentence suggests that while tidy datasets follow a uniform standard , making them structurally similar, messy datasets can be disorganized in various unique ways. The second sentence talks about how it's usually clear which data are variables and which are obsersvations based on their structure or nature, but understanding what these variables and observations in a greater context or outside of a table setting is much more difficult.

  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  
  Wickham defines values as the individual measurements that compose a dataset, variables as all values measuring the same underlying attribute across units, and observations as all values measured on the same unit across attributes.

  5. How is "Tidy Data" defined in section 2.3?
  
  "Tidy data" is defined as data arranged such that each variable forms a column, each observation forms a row, and each type of observational unit forms a table.

  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?

  The five most common problems are: column headers as values not names, multiple variables stored in one column, variables stored in both rows and columns, multiple types of observational units in the same table, and a single observational unit spread across multiple tables. Table 4 is messy because its column headers represent values (income ranges) rather than variable names. Melting a dataset involves turning columns into rows, making the dataset "longer".

  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  
  Table 11 is messy because it stores variables in both rows and columns, while Table 12 is tidy and "molten" because it reorganizes the data so each column represents a variable and each row an observation.

  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

  The "chicken-and-egg" problem refers to the challenge of focusing on tidy data when the tools for data analysis themselves may not support tidy data principles. Wickham hopes that future work will develop more tidy tools, making data analysis easier and more intuitive, which should encourage the creation and use of tidy datasets.


**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [1]:
! git clone https://github.com/DS3001/wrangling

Cloning into 'wrangling'...
remote: Enumerating objects: 83, done.
remote: Counting objects: 100% (40/40), done.
remote: Compressing objects: 100% (29/29), done.
remote: Total 83 (delta 31), reused 11 (delta 11), pack-reused 43
Receiving objects: 100% (83/83), 10.85 MiB | 20.46 MiB/s, done.
Resolving deltas: 100% (36/36), done.


In [17]:
import pandas as pd
import numpy as np

df = pd.read_csv('./wrangling/assignment/data/airbnb_hw.csv', low_memory = False)
price = df['Price']

In [14]:
price = price.str.replace(',', '') # Remove commas from prices
price = pd.to_numeric(price, errors='coerce')
print('The number of missing prices is: ', sum(price.isnull()))


The number of missing prices is:  0


For the prices variable, I cleaned the data by removing the commas from the numbers and converting them to numeric values.

In [16]:
print(price.unique())

[  145    37    28   199   549   149   250    90   270   290   170    59
    49    68   285    75   100   150   700   125   175    40    89    95
    99   499   120    79   110   180   143   230   350   135    85    60
    70    55    44   200   165   115    74    84   129    50   185    80
   190   140    45    65   225   600   109  1990    73   240    72   105
   155   160    42   132   117   295   280   159   107    69   239   220
   399   130   375   585   275   139   260    35   133   300   289   179
    98   195    29    27    39   249   192   142   169  1000   131   138
   113   122   329   101   475   238   272   308   126   235   315   248
   128    56   207   450   215   210   385   445   136   247   118    77
    76    92   198   205   299   222   245   104   153   349   114   320
   292   226   420   500   325   307    78   265   108   123   189    32
    58    86   219   800   335    63   229   425    67    87  1200   158
   650   234   310   695   400   166   119    62   

In [21]:
sdf = pd.read_csv('./wrangling/assignment/data/sharks.csv', low_memory=False)
type = sdf['Type']
type.value_counts()

<ipython-input-21-1708df4f6e46>:1: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  sdf = pd.read_csv('./wrangling/assignment/data/sharks.csv')


In [25]:
type = type.replace(['Boating', 'Sea Disaster', 'Boat', 'Boatomg'], 'Watercraft')
type = type.replace(['Invalid', 'Questionable', 'Unconfirmed', 'Unverified', 'Under investigation'], np.nan)
type.value_counts()
sdf['Type'] = type

Unprovoked    4716
Provoked       593
Watercraft     583
Name: Type, dtype: int64

I cleaned the type variable in the shark dataset by combining the water-based incidents into one category and removing all invalid data.

In [35]:
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
df = pd.read_csv(url,low_memory=False)

In [38]:
released = df['WhetherDefendantWasReleasedPretrial']
print(released.value_counts())
released = released.replace(9, np.nan)
print(released.value_counts())
df['WhetherDefendantWasReleasedPretrial'] = released

1    19154
0     3801
9       31
Name: WhetherDefendantWasReleasedPretrial, dtype: int64
1.0    19154
0.0     3801
Name: WhetherDefendantWasReleasedPretrial, dtype: int64


In [44]:
length = df['ImposedSentenceAllChargeInContactEvent']
type = df['SentenceTypeAllChargesAtConvictionInContactEvent']

length = pd.to_numeric(length, errors='coerce')
length_NA = length.isnull()
print( pd.crosstab(length_NA, type), '\n')
print( np.sum(length_NA))

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274 

9053


In [45]:
length = length.mask( type == 4, 0)
length = length.mask( type == 9, np.nan)
length_NA = length.isnull()
print( pd.crosstab(length_NA, type), '\n')
print( np.sum(length_NA), '\n')
df['ImposedSentenceAllChargeInContactEvent'] = length

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914  8779    0
True                                                 0     0    0     0  274 

274 



Cleaned the data by reducing missing data from 9053 to 274, done through data conversion, missing dummies, and replacing values depending on sentence category.


**Q3.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?